gas-service/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+                gas_token,
+                gas_fee_amount,
+                refund_address,
+            },
+        );
+    }
+
+    #[payable("EGLD")]
+    #[endpoint(addNativeExpressGas)]
+    fn add_native_express_gas(
+        &self,
+        tx_hash: ManagedBuffer,
+        log_index: BigUint,
+        refund_address: ManagedAddress,
+    ) {
+        let value = self.call_value().egld_value().clone_value();
+
+        require!(value > 0, "Nothing received");
+
+        self.native_express_gas_added_event(
+            tx_hash,
+            log_index,
+            AddNativeGasData {
+                value,
+                refund_address,
+            },
+        );
+    }
+
+    #[endpoint(collectFees)]
+    fn collect_fees(
+        &self,
+        receiver: &ManagedAddress,
+        tokens: MultiValueManagedVecCounted<EgldOrEsdtTokenIdentifier>,
+        amounts: MultiValueManagedVecCounted<BigUint>,
+    ) {
+        self.require_only_collector();
+
+        require!(!receiver.is_zero(), "Invalid address");
+
+        let tokens_length = tokens.len();
+        require!(tokens_length == amounts.len(), "Invalid amounts");
+
+        let tokens_vec = tokens.into_vec();
+        let amounts_vec = amounts.into_vec();
+
+        let mut payments = ManagedVec::new();
+
+        for index in 0..tokens_length {
+            let token: EgldOrEsdtTokenIdentifier = tokens_vec.get(index);
+            let amount = amounts_vec.get(index).clone_value();
+
+            require!(amount > 0, "Invalid amounts");
+
+            let balance = self.blockchain().get_sc_balance(&token, 0);
+
+            if token.is_egld() {
+                if amount <= balance {
+                    self.send().direct_egld(receiver, &amount);
+                }
+            } else if amount <= balance {
+                payments.push(EsdtTokenPayment::new(token.unwrap_esdt(), 0, amount));
+            }
+        }
+
+        if !payments.is_empty() {
+            self.send().direct_multi(receiver, &payments);
+        }
+    }
+
+    #[endpoint(refund)]
+    fn refund(
+        &self,
+        tx_hash: ManagedBuffer,
+        log_index: BigUint,
+        receiver: ManagedAddress,
+        token: EgldOrEsdtTokenIdentifier,
+        amount: BigUint,
+    ) {
+        self.require_only_collector();
+
+        require!(!receiver.is_zero(), "Invalid address");
+
+        self.send().direct(&receiver, &token, 0, &amount);
+
+        self.refunded_event(
+            tx_hash,
+            log_index,
+            RefundedData {
+                receiver,
+                token,
+                amount,
+            },
+        );
+    }
+
+    fn require_only_collector(&self) {
+        let caller = self.blockchain().get_caller();
+        let collector = self.gas_collector().get();
+
+        require!(caller == collector, "Not collector");
+    }
+
+    #[view]
+    #[storage_mapper("gas_collector")]
+    fn gas_collector(&self) -> SingleValueMapper<ManagedAddress>;
```

Comment:
There is a need for a function that sets this storage in case the `gas_collector` needs to be changed.

Reply:
I have added such function. Although in the reference implementation the `gas_collector` is immutable and can never be changed, since the `GasService` contract is controlled by the Relayer Microservice, I don't see any reason to not be able to change it here.